plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -231,7 +231,7 @@ public void reset() {
 
     // package-private for tests
     boolean isEof() {
-        return eof;
+        return eof | start + currentOffset == currentStreamLastOffset;
```

Comment:
Probably better to stick with `||` out of principle even if the right hand side is cheap here :)